implementation/src/main/java/io/smallrye/jwt/auth/principal/DecryptionKeyLocationResolver.java

```diff
@@ -0,0 +1,202 @@
+        } catch (Exception e) {
+            if (authContextInfo.getPublicKeyContent() != null) {
+                throw PrincipalMessages.msg.failedToLoadKey(e);
+            } else {
+                throw PrincipalMessages.msg.failedToLoadKeyFromLocation(authContextInfo.getDecryptionKeyLocation(), e);
+            }
+        }
+    }
+
+    @Override
+    public Key resolveKey(JsonWebEncryption jwe, List<JsonWebStructure> nestingContext)
+            throws UnresolvableKeyException {
+        verifyKid(jwe, authContextInfo.getTokenDecryptionKeyId());
+
+        // The verificationKey may have been calculated in the constructor from the local PEM, or,
+        // if authContextInfo.getTokenKeyId() is not null - from the local JWK(S) content.
+        if (decryptionKey != null) {
+            return decryptionKey;
+        }
+
+        // At this point the key can be loaded from either the HTTPS or local JWK(s) content using
+        // the current token kid to select the key.
+        PrivateKey key = tryAsJwk(jwe);
+
+        if (key == null) {
+            if (authContextInfo.getDecryptionKeyContent() != null) {
+                throw PrincipalMessages.msg.failedToLoadKeyWhileResolving();
+            } else {
+                throw PrincipalMessages.msg
+                        .failedToLoadKeyFromLocationWhileResolving(authContextInfo.getDecryptionKeyLocation());
+            }
+        }
+        return key;
+    }
+
+    private PrivateKey tryAsJwk(JsonWebEncryption jwe) throws UnresolvableKeyException {
+        String kid = getKid(jwe);
+
+        if (httpsJwks != null) {
+            return getHttpsJwk(kid);
+        } else if (jsonWebKeys != null) {
+            return getJsonWebKey(kid, jsonWebKeys);
+        } else {
+            return null;
+        }
+    }
+
+    PrivateKey getHttpsJwk(String kid) {
+        PrincipalLogging.log.tryCreateKeyFromHttpsJWKS();
+
+        try {
+            List<JsonWebKey> theKeys = httpsJwks.getJsonWebKeys();
+            PrivateKey theKey = getJsonWebKey(kid, theKeys);
+            if (theKey != null || isMatchingJwkAvailable(theKeys, kid)) {
+                return theKey;
+            }
+        } catch (Exception e) {
+            PrincipalLogging.log.failedToCreateKeyFromJWKSet(e);
+        }
+
+        forcedHttpsJwksRefresh();
+
+        try {
+            PrincipalLogging.log.tryCreateKeyFromJWKSAfterRefresh();
+            return getJsonWebKey(kid, httpsJwks.getJsonWebKeys());
+        } catch (Exception e) {
+            PrincipalLogging.log.failedToCreateKeyFromJWKSAfterRefresh(e);
+        }
+        return null;
+    }
+
+    PrivateKey getJsonWebKey(String kid, List<JsonWebKey> jwks) {
+        PrincipalLogging.log.tryCreateKeyFromJWKS();
+
+        try {
+            JsonWebKey jwk = getJsonWebKey(kid, jwks, authContextInfo.getKeyEncryptionAlgorithm().getAlgorithm());
+            return jwk == null ? null : PublicJsonWebKey.class.cast(jwk).getPrivateKey();
```

Comment:
Is `PublicJsonWebKey` the correct cast? Just verifying.

Reply:
Yes, Jose4J uses `PublicJsonWebKey` to capture both public and private keys, I'll add a test

Reply:
Hmm...Need to look deeper

Reply:
It is all good, the test is working fine